File: apps/cargo-tracker/src/contexts/booking/domain/model/cargo-itinerary.ts
import { Location } from '../../../../shared/domain/model/location.js';
import { BookingValidationError } from './booking-validation-error.js';

interface LegProps {
  voyageNumber: string;
  loadLocation: string;
  unloadLocation: string;
  loadTime: Date;
  unloadTime: Date;
}

/**
 * 輸送区間（Booking Context 値オブジェクト）。単一航海での積込港から荷降港までの区間。
 * VoyageNumber は Routing Context 固有型のため、Booking では文字列として保持する（BC 独立性）。
 */
export class Leg {
  private constructor(
    readonly voyageNumber: string,
    readonly loadLocation: Location,
    readonly unloadLocation: Location,
    readonly loadTime: Date,
    readonly unloadTime: Date,
  ) {}

  static of(props: LegProps): Leg {
    const voyageNumber = props.voyageNumber.trim();
    if (voyageNumber.length === 0) {
      throw new BookingValidationError('航海番号は必須です');
    }
    let load: Location;
    let unload: Location;
    try {
      load = Location.of(props.loadLocation);
      unload = Location.of(props.unloadLocation);
    } catch (error) {
      throw new BookingValidationError(
        error instanceof Error ? error.message : '積込地・荷降地の指定が不正です',
      );
    }
    if (load.sameAs(unload)) {
      throw new BookingValidationError('積込地と荷降地は異なる必要があります');
    }
    if (props.loadTime.getTime() > props.unloadTime.getTime()) {
      throw new BookingValidationError('積込時刻は荷降時刻以前である必要があります');
    }
    return new Leg(voyageNumber, load, unload, props.loadTime, props.unloadTime);
  }
}

/** 旅程（Booking Context 値オブジェクト）。1 つ以上の Leg で構成し、連結制約を満たす */
export class CargoItinerary {
  private constructor(readonly legs: Leg[]) {}

  static of(legs: Leg[]): CargoItinerary {
    if (legs.length === 0) {
      throw new BookingValidationError('旅程は 1 つ以上の輸送区間で構成される必要があります');
    }
    for (let i = 0; i < legs.length - 1; i += 1) {
      if (!legs[i].unloadLocation.sameAs(legs[i + 1].loadLocation)) {
        throw new BookingValidationError(
          `輸送区間が連結していません（${legs[i].unloadLocation.unlocode} ≠ ${legs[i + 1].loadLocation.unlocode}）`,
        );
      }
    }
    return new CargoItinerary([...legs]);
  }

  expectedArrivalTime(): Date {
    return this.legs[this.legs.length - 1].unloadTime;
  }
}

File: apps/backend/bookingms/src/main/java/com/example/bookingms/domain/model/valueobjects/CargoItinerary.java
package com.example.bookingms.domain.model.valueobjects;

import com.example.shared.domain.model.Location;
import java.time.Instant;
import java.util.List;

/**
 * 旅程。予約に割り当てられた、積み替えを含む全区間（US09）。
 *
 * <p>Routing Context の {@code TransitPath} と<strong>同じ不変条件を、別の型として持つ</strong>。
 * あちらは都度算出して捨てる探索結果であり、こちらは予約に紐付いて残る記録である。共有すると、
 * 探索の都合（推奨順・費用の概算）が予約側の記録に混ざる。
 */
public final class CargoItinerary {

    private final List<Leg> legs;

    private CargoItinerary(List<Leg> legs) {
        this.legs = List.copyOf(legs);
    }

    /**
     * 新規に組み立てる。ここでだけ検査する。
     *
     * <p>つながっていない旅程を保存すると、荷役の担当者は来ない貨物を待つことになる。
     */
    public static CargoItinerary of(List<Leg> legs) {
        if (legs == null || legs.isEmpty()) {
            throw new IllegalArgumentException("旅程には少なくとも 1 つの区間が必要です");
        }
        for (int i = 1; i < legs.size(); i++) {
            Leg previous = legs.get(i - 1);
            Leg current = legs.get(i);
            if (!previous.unloadLocation().equals(current.loadLocation())) {
                throw new IllegalArgumentException(
                        "区間がつながっていません。前の区間の荷降し地から次の区間が積み込むようにしてください");
            }
            if (current.loadTime().isBefore(previous.unloadTime())) {
                throw new IllegalArgumentException("次の区間の積込は前の区間の荷降しより後にしてください");
            }
        }
        return new CargoItinerary(legs);
    }

    /**
     * 永続化された行から復元する。ここでは検査しない。
     *
     * <p>連結の規則が無かったころの行が読めなくなると、一覧そのものが開けなくなる。
     */
    public static CargoItinerary restore(List<Leg> legs) {
        return new CargoItinerary(legs);
    }

    public List<Leg> legs() {
        return legs;
    }

    public Location origin() {
        return legs.get(0).loadLocation();
    }

    public Location destination() {
        return legs.get(legs.size() - 1).unloadLocation();
    }

    /** 出発予定。最初の区間の積込時刻。 */
    public Instant expectedDepartureTime() {
        return legs.get(0).loadTime();
    }

    /** 到着予定。最後の区間の荷降し時刻。 */
    public Instant expectedArrivalTime() {
        return legs.get(legs.size() - 1).unloadTime();
    }

    /**
     * その地点を通る予定か。
     *
     * <p>誤配の判定（US28）は「そこを通る予定だったか」を旅程に聞く。荷役の場所と突き合わせる
     * のを呼ぶ側の仕事にすると、入口ごとに別の判定が書かれる。
     */
    public boolean includesLocation(Location location) {
        return legs.stream().anyMatch(leg -> leg.loadLocation().equals(location)
                || leg.unloadLocation().equals(location));
    }

    /** 積み替えの回数。区間の数から 1 を引いたもの。 */
    public int transshipmentCount() {
        return legs.size() - 1;
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof CargoItinerary itinerary && legs.equals(itinerary.legs);
    }

    @Override
    public int hashCode() {
        return legs.hashCode();
    }

    @Override
    public String toString() {
        return "%s → %s（%d 区間）".formatted(origin().unLocode(), destination().unLocode(),
                legs.size());
    }
}
